### PR TITLE
Finish halfbrained's awesome work

### PR DESCRIPTION
**Pull request by Akuli**
- [x] Figure out when hover responses can be null, add test: created #45

### CODE REVIEW

File: sansio_lsp_client/structs.py [hunk 1]
```diff
@@ -51,11 +54,15 @@ class VersionedTextDocumentIdentifier(TextDocumentIdentifier):
     version: t.Optional[int]
 
 
+# Sorting tip:  sorted(positions, key=(lambda p: (p.line, p.character)))
```
Comment (Akuli) [inline sansio_lsp_client/structs.py:57]:
```suggestion
# Sorting tip:  sorted(positions, key=(lambda p: p.as_tuple()))
```

File: sansio_lsp_client/structs.py [hunk 2]
```diff
@@ -203,19 +260,189 @@ class DiagnosticSeverity(enum.IntEnum):
...
     HINT = 4
 
+    # TODO: delete this? it's easy to implement in your editor if you really need it
+    def short_name(self) -> str:
+        return {
+            self.ERROR: "Err",
+            self.WARNING: "Wrn",
+            self.INFORMATION: "Inf",
+            self.HINT: "Hint",
+        }[self]
```
Comment (Akuli) [inline sansio_lsp_client/structs.py:270]: @halfbrained What do you think about this?

File: sansio_lsp_client/structs.py [hunk 3]
```diff
@@ -51,11 +54,15 @@ class VersionedTextDocumentIdentifier(TextDocumentIdentifier):
...
 
 
+# Sorting tip:  sorted(positions, key=(lambda p: (p.line, p.character)))
 class Position(BaseModel):
     # NB: These are both zero-based.
     line: int
     character: int
 
+    def as_tuple(self) -> t.Tuple[int, int]:
+        return (self.line, self.character)
```
Comment (Akuli) [inline sansio_lsp_client/structs.py:64]: @halfbrained My `as_tuple()` method replaces `__lt__` in this class and `pos()` in `class DocumentSymbol`. What do you think?

File: sansio_lsp_client/structs.py [hunk 4]
```diff
@@ -192,6 +242,13 @@ class Location(BaseModel):
     range: Range
 
 
+class LocationLink(BaseModel):
+    originSelectionRange: t.Optional[Range]
+    targetUri: str  # DocumentUri...
```
Comment (Akuli) [inline sansio_lsp_client/structs.py:247]: @halfbrained What does this comment mean?

File: sansio_lsp_client/structs.py [hunk 5]
```diff
@@ -203,19 +260,189 @@ class DiagnosticSeverity(enum.IntEnum):
...
+class SymbolInformation(BaseModel):  # symbols: flat list
+    name: str
+    kind: SymbolKind
+    tags: t.Optional[t.List[SymbolTag]]
+    deprecated: t.Optional[bool]
+    location: Location
+    containerName: t.Optional[str]
+
+
+class DocumentSymbol(BaseModel):  # symbols: hierarchy
```
Comment (Akuli) [inline sansio_lsp_client/structs.py:377]: @halfbrained What do the `# symbols: ...` comments mean?

File: tests/test_actual_langservers.py [hunk 6]
```diff
@@ -0,0 +1,551 @@
...
+
+        self.reader_thread.start()
+        self.writer_thread.start()
+
+        self.exception = None
+
+    # threaded
+    def _read_loop(self):
+        try:
+            while self._pout:
```
Comment (Akuli) [inline tests/test_actual_langservers.py:102]:
```suggestion
            while True:
```

File: tests/test_actual_langservers.py [hunk 7]
```diff
@@ -0,0 +1,551 @@
...
+
+                self._read_q.put(data)
+        except Exception as ex:
+            self.exception = ex
+        self._send_q.put_nowait(None)  # stop send loop
+
+    # threaded
+    def _send_loop(self):
+        try:
+            while self._pin:
```
Comment (Akuli) [inline tests/test_actual_langservers.py:116]:
```suggestion
            while True:
```

File: tests/test_actual_langservers.py [hunk 8]
```diff
@@ -0,0 +1,551 @@
...
+
+        assert self.lsp_client.is_initialized
+        self.lsp_client.shutdown()
+        self.wait_for_message_of_type(lsp.Shutdown)
+        self.lsp_client.exit()
+        self._queue_data_to_send()
+        self._read_data_received()
+
+    def do_method(self, text, file_uri, method, response_type=None):
+        def doc_pos():  # SKIP
```
Comment (Akuli) [inline tests/test_actual_langservers.py:196]: @halfbrained What doees this comment mean?

File: sansio_lsp_client/structs.py (hunk 3 above)
Comment (Akuli) [reply sansio_lsp_client/structs.py]: After making this review comment, I changed the lambda to `lambda p: p.as_tuple()`, which is even simpler.

File: sansio_lsp_client/structs.py (hunk 2 above)
Comment (halfbrained) [reply sansio_lsp_client/structs.py:270]: No objections

File: sansio_lsp_client/structs.py (hunk 3 above)
Comment (halfbrained) [reply sansio_lsp_client/structs.py]: Is there a benefit to this change? :)

File: sansio_lsp_client/structs.py (hunk 4 above)
Comment (halfbrained) [reply sansio_lsp_client/structs.py:247]: Just a note that in the specification this field is of type `DocumentUri`

File: sansio_lsp_client/structs.py (hunk 5 above)
Comment (halfbrained) [reply sansio_lsp_client/structs.py:377]: Meaning is: `SymbolInformation` is just a flat list of symbols, `DocumentSymbol` - a tree - `METHOD` items are children of `CLASS` items, etc. 

It's not clear from just the type names.

File: tests/test_actual_langservers.py (hunk 8 above)
Comment (halfbrained) [reply tests/test_actual_langservers.py:195]: It's for my editor :) no other purpose

File: sansio_lsp_client/structs.py (hunk 4 above)
Comment (Akuli) [reply sansio_lsp_client/structs.py:247]:
```suggestion
    targetUri: str  # in the spec the type is DocumentUri
```

File: sansio_lsp_client/structs.py (hunk 3 above)
Comment (Akuli) [reply sansio_lsp_client/structs.py]: I think `Position.__lt__()` and `DocumentSymbol.pos()` were both weird:
- If there is `__lt__`, why are `__gt__`, `__le__` and `__ge__` missing? It's because `__lt__` is what `sort()` and `sorted()` use, but it's not obvious when looking at the code. We could add a comment about it, but `key=(lambda p: p.as_tuple())` is very obvious as is.
- If you see `something.pos()` in the middle of longer code, you wouldn't expect it to take just the **start** of a range and ignore the end, which is what that method did. On the other hand, if you see `something.selectionRange.start.as_tuple()`, you immediately know it uses just the start position and the end is being ignored.

**Comment by Akuli** [comment]
Marking as draft because I split some of this to separate pull requests that should be merged first.

**Comment by PurpleMyst** [comment]
Hey @Akuli, I saw you've requested my review on a couple PRs. I've been *very* busy with an upcoming exam. I'll be free in the incoming week, so I'll take a look at them then. Sorry for the delay.

**Comment by Akuli** [comment]
Awesome! Let's review the other PRs next week.

File: tests/test_actual_langservers.py [hunk 9]
```diff
@@ -106,12 +132,26 @@ def _read_data_received(self):
...
+            msg.reply([lsp.WorkspaceFolder(uri=self.root_uri, name="Root")])
+
+    #        else:
+    #            print(f"Can't autoreply: {type(msg)}")
+
     def wait_for_message_of_type(self, type_, timeout=5):
-        end_time = time.monotonic() + timeout
-        while time.monotonic() < end_time:
+        end_time = time.time() + timeout
+        while time.time() < end_time:
```
Comment (Akuli) [inline tests/test_actual_langservers.py:154]: Merge problem

```suggestion
        end_time = time.monotonic() + timeout
        while time.monotonic() < end_time:
```

File: tests/test_actual_langservers.py [hunk 10]
```diff
@@ -322,8 +379,7 @@ def do_bar():
             assert diag_msgs == [
                 "Non-void function does not return a value",
                 "Use of undeclared identifier 'do_'",
-                "Expected '}'\n\nfoo.c:9:16: note: to match this '{'",
-                "To match this '{'\n\nfoo.c:9:37: error: expected '}'",
+                "Expected '}'",
```
Comment (Akuli) [inline tests/test_actual_langservers.py:382]: No idea why this changes

File: tests/test_actual_langservers.py [hunk 11]
```diff
@@ -349,6 +406,101 @@ def do_bar():
...
+            assert len(typedef.result) == 1
+            assert typedef.result[0].uri == (project_root / filename).as_uri()
+
+
+def _needs_clangd(version):
+    return pytest.mark.skipif(
+        not list(langserver_dir.glob(f"clangd_{version}.*")),
+        reason=f"clangd {version} not found",
+    )
+
```
Comment (Akuli) [inline tests/test_actual_langservers.py:504]: Not used

```suggestion
```

**Comment by Akuli** [comment]
@PurpleMyst Ready for review

File: sansio_lsp_client/client.py [hunk 12]
```diff
@@ -171,14 +329,49 @@ def parse_request(event_cls: t.Type[Event]) -> Event:
...
         elif request.method == "window/logMessage":
             return parse_request(LogMessage)
+
         elif request.method == "textDocument/publishDiagnostics":
             return parse_request(PublishDiagnostics)
+
+        elif request.method == "window/workDoneProgress/create":
+            assert request.params is not None
+            event = parse_request(WorkDoneProgressCreate)
+            self._progress_tokens_map[request.params["token"]] = WorkDoneProgress
```
Comment (PurpleMyst) [inline sansio_lsp_client/client.py:351]: Why is this a `dict`? It seems to only have `WorkDoneProgress` as a value

File: sansio_lsp_client/client.py [hunk 13]
```diff
@@ -211,7 +410,7 @@ def shutdown(self) -> None:
 
     def exit(self) -> None:
         assert self._state == ClientState.SHUTDOWN
-        self._send_notification(method="exit")
+        self._send_notification(method="exit", params={})
```
Comment (PurpleMyst) [inline sansio_lsp_client/client.py:413]: I'd like it if we could put a comment here explaining why we can't just not specify `params`

File: sansio_lsp_client/client.py [hunk 14]
```diff
@@ -285,3 +495,111 @@ def completions(
...
+    def signatureHelp(self, text_document_position: TextDocumentPosition) -> int:
+        assert self._state == ClientState.NORMAL
+        return self._send_request(
+            method="textDocument/signatureHelp", params=text_document_position.dict()
+        )
+
+    def definition(
+        self,
+        text_document_position: TextDocumentPosition,
+        # TODO PartialResultParams
```
Comment (PurpleMyst) [inline sansio_lsp_client/client.py:514]: What does this comment mean?

File: sansio_lsp_client/events.py [hunk 15]
```diff
@@ -78,3 +130,98 @@ class WillSaveWaitUntilEdits(Event):
 class PublishDiagnostics(ServerNotification):
     uri: str
     diagnostics: t.List[Diagnostic]
+
+
+class Hover(Event):
```
Comment (PurpleMyst) [inline sansio_lsp_client/events.py:135]: What do the comments in this class body mean?

File: sansio_lsp_client/structs.py [hunk 16]
```diff
@@ -51,11 +54,15 @@ class VersionedTextDocumentIdentifier(TextDocumentIdentifier):
     version: t.Optional[int]
 
 
+# Sorting tip:  sorted(positions, key=(lambda p: p.as_tuple()))
```
Comment (PurpleMyst) [inline sansio_lsp_client/structs.py:57]: Should we add some comparison operators here so that this is redundant?

**Review by PurpleMyst — CHANGES_REQUESTED** [review]
LGTM but I'm still iffy about the whole logging business. Perhaps we could do it a different way? Don't feel great about having a hidden switch just for testing.

File: sansio_lsp_client/structs.py (hunk 16 above)
Comment (Akuli) [reply sansio_lsp_client/structs.py:57]: There used to be comparison operators (actually just `__lt__`), but I switched to this because it seemed a lot simpler. Imagine seeing this in the middle of longer code:
- `sorted(positions)` is like "Can you really do this? Does it work? Is this a typo?"
-  `sorted(positions, key=...)` is like "We compare them as tuples, so it works"

Before `.as_tuple()`, there was also a convenience method that returned `foo.range.start.as_tuple()` (I think), and that was replaced by `.as_tuple()` too. Overall `.as_tuple()` feels like a nice and minimal approach to me.

Comment (PurpleMyst) [reply sansio_lsp_client/structs.py:57]: Understandable.

File: sansio_lsp_client/client.py (hunk 13 above)
Comment (Akuli) [reply sansio_lsp_client/client.py:413]: I thought this was another gopls thing, but changing it to `params=None` doesn't break gopls tests.

File: sansio_lsp_client/client.py (hunk 12 above)
Comment (Akuli) [reply sansio_lsp_client/client.py]: I cleaned it up. I think halfbrained's original idea was to implement more stuff later, but we can easily add this back if it's needed.

**Comment by Akuli** [comment]
I deleted the logging stuff. Its purpose was to allow receiving events even if receiving one of them errors, so that you still get the other events. I changed `recv()` to return a generator that yields correctly parsed messages and then fails, but doesn't leave the failed message to `_recv_buf` so that it will succeed next time. I feel like it's almost too elegant; the code doesn't look tricky, even though I have specifically written it so that it handles many different cases.

**Comment by Akuli** [comment]
@halfbrained, can you explain your comments to us for one last time? :)

File: sansio_lsp_client/events.py (hunk 15 above)
Comment (Akuli) [reply sansio_lsp_client/events.py:135]: They listed attributes that the members had, which is just duplicate information, because you can go to where the member type is defined and look. I deleted them.

File: sansio_lsp_client/client.py (hunk 13 above)
Comment (Akuli) [reply sansio_lsp_client/client.py:378]: I added a TODO comment about this

File: sansio_lsp_client/client.py (hunk 14 above)
Comment (Akuli) [reply sansio_lsp_client/client.py]: I don't know, but if it's something important, we'll discover it later. I deleted the comments for now. They really aren't of any value when nobody understands them.

**Comment by Akuli** [comment]
@PurpleMyst Ready for review again

**Comment by PurpleMyst** [comment]
LGTM!

File: sansio_lsp_client/client.py (hunk 13 above)
Comment (halfbrained) [reply sansio_lsp_client/client.py:378]: It was indeed added in gopls fixes, but I don't remember the events leading up to it, and I already deleted all the go stuff, sorry.

File: tests/test_actual_langservers.py (hunk 10 above)
Comment (halfbrained) [reply tests/test_actual_langservers.py:381]: If I recall correctly - that's a result of sending client's capabilities during the server initialization

**Comment by halfbrained** [comment]
> @halfbrained, can you explain your comments to us for one last time? :)

My apologies for missing this.

(Not sure how that happened, but I'm almost certain that's github having a laugh at my expense... :) )
